File: ios/Sources/ThumbnailPlugin.swift
import UIKit
import WebKit
import QuickLookThumbnailing
import Tauri

class ThumbnailPlugin: Plugin {
    @objc public func getThumbnail(_ invoke: Invoke) {
        let path = invoke.getString("path") ?? ""
        let width = invoke.getInt("width") ?? 200
        let height = invoke.getInt("height") ?? 200
        
        if path.isEmpty {
            invoke.reject("Path is required")
            return
        }
        
        let url = URL(fileURLWithPath: path)
        let size = CGSize(width: CGFloat(width), height: CGFloat(height))
        let scale = UIScreen.main.scale
        
        let request = QLThumbnailGenerator.Request(
            fileAt: url,
            size: size,
            scale: scale,
            representationTypes: .thumbnail
        )
        
        QLThumbnailGenerator.shared.generateBestRepresentation(for: request) { thumbnail, error in
            if let error = error {
                invoke.reject("Error generating thumbnail: \(error.localizedDescription)")
                return
            }
            
            guard let thumbnail = thumbnail else {
                invoke.reject("Failed to generate thumbnail")
                return
            }
            
            let image = thumbnail.uiImage
            if let data = image.pngData() {
                invoke.resolve([
                    "thumbnail": data,
                    "mimeType": "image/png"
                ])
            } else {
                invoke.reject("Failed to encode thumbnail to PNG")
            }
        }
    }
}

@objc public init_plugin_thumbnail() -> Plugin {
  return ThumbnailPlugin()
}

File: android/src/main/java/com/tauri/thumbnail/ThumbnailPlugin.kt
package com.tauri.thumbnail

import android.app.Activity
import android.graphics.Bitmap
import android.media.ThumbnailUtils
import android.util.Size
import android.webkit.MimeTypeMap
import app.tauri.annotation.Command
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.JSObject
import app.tauri.plugin.Plugin
import app.tauri.plugin.Invoke
import java.io.ByteArrayOutputStream
import java.io.File
import android.util.Base64

@TauriPlugin
class ThumbnailPlugin(private val activity: Activity) : app.tauri.plugin.Plugin(activity) {

    @Command
    fun getThumbnail(invoke: Invoke) {
        val args = invoke.getArgs()
        val path = args.getString("path", null)
        val width = args.getInteger("width", 200)
        val height = args.getInteger("height", 200)

        if (path == null) {
            invoke.reject("Path is required")
            return
        }

        val file = File(path)
        if (!file.exists()) {
            invoke.reject("File does not exist: $path")
            return
        }

        try {
            val size = Size(width, height)
            val bitmap = if (isVideo(path)) {
                ThumbnailUtils.createVideoThumbnail(file, size, null)
            } else {
                ThumbnailUtils.createImageThumbnail(file, size, null)
            }

            if (bitmap == null) {
                invoke.reject("Failed to generate thumbnail for $path")
                return
            }

            val outputStream = ByteArrayOutputStream()
            bitmap.compress(Bitmap.CompressFormat.PNG, 100, outputStream)
            val byteArray = outputStream.toByteArray()
            val base64 = Base64.encodeToString(byteArray, Base64.NO_WRAP)

            val response = JSObject()
            response.put("thumbnail", "data:image/png;base64,$base64")
            invoke.resolve(response)
        } catch (e: Exception) {
            invoke.reject("Error generating thumbnail: ${e.message}")
        }
    }

    private fun isVideo(path: String): Boolean {
        val extension = MimeTypeMap.getFileExtensionFromUrl(path)
        val mimeType = MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension.lowercase())
        return mimeType?.startsWith("video/") == true
    }
}
